README.md
# Crystal Maze Games

A web framework for playing Crystal Maze like games. 

## [Lights Out](http://www.crystalmazedatabase.com/game.php?id=2)

In this game, a random set of switches are provided to the user and a light board. Some switches will turn off lights from the light board, ultimately revealing
a 4 digit combination. However, some of the buttons will "reset" the light board. The player must remember which switches reset the board and avoid them to reveal
the code.

In order to play Lights Out, a config file named `lightsout.yaml` must be provided in the config directory. This must contain the following config

1. `answer` - The numerical answer the player needs to guess

## Who's on First

Based on the game Who's on First from [Keep Talking and Nobody Explodes](https://keeptalkinggame.com/). In this game, 1 player is presented with a set of buttons
with text on them. They score points by clicking on the correct buttons. The button to press is given to the second player who describes the button to the first
player.

In order to play Who's on First, a config file named 'whosonfirst.yaml' must be provided in the config directory. This must container the following config

1. `slack_url` - The slack webhook URL to send the picture of the button to press.
1. `scores` - A list of dictionaries which act as a scoreboard. Must be provided in ascending order. Each entry may contain
	1. `score` - Mandatory, the score for this position
	1. `urls` - Optional, a list of urls for the picture of the score (recommended that this looks like a button)
1. `buttons` - A list of buttons to display to the user. Each button can have the following config
	1. `colour` - The colour of the button
	1. `text_colour` - The colour of the text that appears on the button
	1. `url` - A URL that is a picture of the button
	1. `text` - A list of text to display on the button. Each list item is on a new line
	1. `text_size` - An optional value, the text size in pixels to draw the text, default is 16.

Additionally, the following configuration options can be provided

1. `slack_icon` - The icon to display on the slack message. If no slack_icon is provided the default slack icon for the webhook is used.
1. `stats` - The file to write stats to (how long a guess took and whether it was correct or not). If not provided, stats are not recorded.
1. `rounds` - The number of buttons to have guesses for before the game ends. If no rounds config is provided, the game will continue until the max score is reached.
1. `gamelength` - The length of time in seconds before the game ends. If no gamelength config is provided, the game will continue until the max score is reached.
1. `height` - The height of the gameboard to display buttons in. Default is 675 pixels.
1. `width` - The width of the gameboard to display buttons in. Default is 1200 pixels.

An example config file is as follows

```
slack_url: https://hooks.slack.com/services/T1A8GVBNH/7KBF7RXKU/BB2p3u8swyom10fK9mPBWthw
# rounds: 5
gamelength: 60
height: 720
width: 1200
scores:
  - score: 0
    urls: 
      - https://i.imgur.com/hJU5yHG.png
      - https://i.imgur.com/HPljlHt.png
  - score : 1
    urls:
      - https://i.imgur.com/KgpbCZH.png
      - https://i.imgur.com/EJxp03a.png
  - score : 2
    urls:
      - https://i.imgur.com/h7xDmr8.png
      - https://i.imgur.com/MEkkXhb.png
  - score : 3
    urls:
      - https://i.imgur.com/qNE17F9.png
      - https://i.imgur.com/fsswiOV.png
buttons:
  - colour: black
    text_colour: white
    url: https://i.imgur.com/7SsPgdc.png
    text: ONE
  - colour: black
    text_colour: white
    text_size: 14
    url: https://i.imgur.com/gxrWdVV.png
    text: one

```

CHANGELOG.md
# [Version 2.0](https://github.com/AdmiralGT/crystal_maze/releases/tag/2.0):
* New game! Who's on first.
* Lights's Out 4 digit code is now provided in a config file.

# [Version 1.1](https://github.com/AdmiralGT/crystal_maze/releases/tag/1.1):
* User is now allowed to guess the answer and is told if they are correct.

# [Version 1.0](https://github.com/AdmiralGT/crystal_maze/releases/tag/1.1):
* A 4 digit code configured on the server (currently hard-coded)
* A predetermined reveal pattern for each digit 

scripts/whosonfirst.js
// SVG Namespace
var svgNS = "http://www.w3.org/2000/svg";

// Variables for the total SVG area
var svg_width = 1200;
var svg_height = 675

var scoreboard_svg_inner = 56;
var scoreboard_svg_border = 4;
var scoreboard_svg_size = scoreboard_svg_inner + scoreboard_svg_border;

// A list of all the buttons and all the locations
var global_button_list = new Array();
var global_button_locations = new Array();

// A list of buttons that are the targets this game
var target_button_list = new Array();

// A list of the buttons in this game
var game_button_list = new Array();

// A list of possible scores
var scores_list = new Array();

// Variables for the buttons that can be pressed
var button_pressed = "white";
var button_radius = 50;
var button_diameter = 2 * button_radius;
var button_separation = 2;
var button_effective_diameter = button_diameter + button_separation;
var score_pos = 0;
var rounds = 0;
var game_in_progress = false;

// Variable to track the timer.
var game_end_timer;
var game_length = -1;

var image_shown = 0; 

// Function called when a button is pressed
// If this is a reset button, reset the light board, otherwise turn off the segment of lights.
function button_press()
{
	// Don't do anything if the game is not in progress, it's just some buttons on a screen.
	if (game_in_progress)
	{
	    var clicked_button = getButtonFromElementID(this.event.target.id);
	    var correct = false;
	    if (clicked_button.target)
	    {
	    	alert("Correct")
	    	correct = true;
	    	set_score_position(score_pos + 1)
	    }
	    else
	    {
	    	alert("Wrong")
	    	set_score_position(score_pos - 1)
	    }

	    var time = new Date().getTime() - image_show
	    send_message_to_server("record_stats?time=" + time + "&correct=" + correct)
    	if (score_pos == scores_list.length - 1)
    	{
    		end_game()
    		return
    	}
	    // We don't know which button is the current target so just set them all not to be the target
	    for (var ii = 0; ii < game_button_list.length; ii++)
	    {
	    	var button = game_button_list[ii]
	    	button.setTarget(false)
	    }

	    determine_next_action()
	}
}

function determine_next_action()
{
    if (rounds > 0)
    {
	    if (target_button_list.length > 0)
	    {
		    set_next_target(target_button_list.pop())
	    }
	    else
	    {
	    	end_game()
	    }
    }
    else
    {
    	if (target_button_list.length == 0)
    	{
		    shuffle(game_button_list)
		    generate_target_list(rounds)
    	}
	    set_next_target(target_button_list.pop())
    }

}

// Function to end the game and print the score
function end_game()
{
   	game_in_progress = false
   	send_message_to_server("post_score?score_pos=" + score_pos);
   	reset_game()
   	alert("Score: " + scores_list[score_pos])
}


// Function to get the next target button and send the image to slack
function set_next_target(button)
{
    button.setTarget(true)
    send_message_to_server("post_whosonfirst_button?imageurl=" + button.imageurl);
    image_show = new Date().getTime();

}

// Function that receives game configuration
function receive_game_config(config)
{
	var json = JSON.parse(config)
	if (json['rounds'])
		rounds = json['rounds']
	if (json['gamelength'])
		game_length = json['gamelength']
  	if (json['width'])
  		svg_width = json['width']
  	if (json['height'])
  	  	svg_height = json['height']
  	size_game_board("svg_area", svg_width, svg_height)

  	if (json['buttons'])
  		generate_buttons(json['buttons'])
 	if (json['scores'])
 	{
 		var scores_dict = json['scores']
 		scores_list = []
 		for (var ii = 0; ii < scores_dict.length; ii++)
 		{
 			if (!isNaN(scores_dict[ii]['score']))
 				scores_list.push(scores_dict[ii]['score'])	
 		}
		 		
  		generate_score_board(scores_list)
  		size_game_board("score_svg", scoreboard_svg_size * scores_list.length, scoreboard_svg_size)
  	}
}

// Function that generates all our necessary buttons
function generate_buttons(buttons)
{
	horizontal = Math.floor(svg_width / button_effective_diameter)
	vertical = Math.floor(svg_height / button_effective_diameter)

	for (var ii = 0; ii < buttons.length; ii++)
    {
    	var button = buttons[ii]
    	if (!Array.isArray(button.text))
    	{
    		button.text = [button.text]
    	}
        make_button(button);
    }

    // Generate the locations for the buttons
    for (var button_num = 0; button_num < buttons.length; button_num++)
    {
      button_location = generate_button_location(button_num, horizontal, vertical)
      global_button_locations.push(button_location);
    }
}

// Function to generate a button location.
function generate_button_location(id, width, height)
{
	y_pos = Math.floor(id / width)
	x_pos = id % width

	x = button_radius + (x_pos * button_effective_diameter)
	y = button_radius + (y_pos * button_effective_diameter)
	var pos = new Position(x,y);
    return pos;
}

// Used on page load to generate the light grid and buttons that will be in game
function generate_game()
{
    get_game_config();
    generate_objects();
}

// Function to generate the guess box and guess button.
function generate_objects()
{
    var gameboard = document.getElementById('scoreboard');
    var div = document.createElement('div');
    div.setAttribute('id', 'guess_div');

    add_start_button(div)
    gameboard.appendChild(div);
}

function add_start_button(div)
{
    var start_button = document.createElement('button');
    start_button.innerHTML = "Start";
    start_button.setAttribute('onclick', 'start_game()');
    start_button.setAttribute('id', 'start_button');
    div.appendChild(start_button);
}

// Get game configuration
function get_game_config()
{
	var button_request = new XMLHttpRequest();
	button_request.open('GET', 'whosonfirst_config', true);
	button_request.onload = function (e) {
		receive_game_config(button_request.responseText)
	}
	button_request.send()
}

// Someone is attempting to guess the answer.
function start_game()
{
	if (!game_in_progress)
	{
		var start_button = document.getElementById("start_button")
		start_button.parentNode.removeChild(start_button);

		set_score_position(0)
		game_in_progress = true;

	    // We can only display a certain number of buttons so slice the list at the max size
	    shuffle(global_button_list)
	    game_button_list = global_button_list.slice(0, horizontal * vertical)
	  	generate_button_grid(game_button_list)
	    set_button_locations(game_button_list)
	    shuffle(game_button_list)
	    generate_target_list(rounds)
	    set_next_target(target_button_list.pop())

	    if (game_length > 0)
	    {
		    game_end_timer = setTimeout(end_game, game_length * 1000)
	    }
	}
}

function generate_target_list(num_rounds)
{
    if (rounds != 0)
    {
    	target_button_list = game_button_list.slice(0, rounds)
    }
	else
	{
		target_button_list = game_button_list.slice()
	}
}

function reset_game()
{
	clearTimeout(game_end_timer)
    var div = document.getElementById('guess_div');
	add_start_button(div)
	var svg_area = document.getElementById("svg_area")
	while (svg_area.firstChild)
	{
		svg_area.firstChild.remove()
	}
	var scoreboard_svg = document.getElementById("score_svg")
	while (scoreboard_svg.firstChild)
	{
		scoreboard_svg.firstChild.remove()
	}
	game_in_progress = false
	global_button_list.length = 0
	global_button_locations.length = 0
	game_button_list.length = 0
	get_game_config()
}

// Sends a picture of the button to slack
function send_message_to_server(url)
{
	var slack_request = new XMLHttpRequest();
	slack_request.open("GET", url, true);
	slack_request.send()
}

// Function to build a button and put it in the global list of buttons.
function make_button(button_json)
{
    //var button = new Button(colour, text, text_colour, button_radius);
    var button = new Button(button_json.colour, button_json.text, button_json.text_colour, button_radius)
    if (button_json.text_size)
    {
	    button.setTextSize(button_json.text_size)
    }
    else
    {
    	button.setTextSize(16)
    }
    if (button_json.url)
    {
	    button.setImageURL(button_json.url)
    }
    else
    {
    	button.setImageURL('https://i.imgur.com/JMy9n3p.jpg?1')
    }
    global_button_list.push(button)
}

// Generate the score board where we display what score the player has
function generate_score_board(score_list)
{
	// Generate the general SVG area to put buttons in.
	var scoreboard_svg = document.getElementById("score_svg");

	for (var score_num = 0; score_num < score_list.length; score_num++)
	{
		var score = score_list[score_num];

		var g = document.createElementNS(svgNS, "g");
		scoreboard_svg.appendChild(g);

		var score_element = document.createElementNS(svgNS, "rect");
		var score_id = generate_id_of_object("score_", score_num, score_list.length)

		score_element.setAttributeNS(null, 'id', score_id);
		score_element.setAttributeNS(null, 'width', scoreboard_svg_inner)
		score_element.setAttributeNS(null, 'height', scoreboard_svg_inner)
		score_element.setAttributeNS(null, 'x', scoreboard_svg_size * score_num + (scoreboard_svg_border / 2))
		score_element.setAttributeNS(null, 'y', scoreboard_svg_border / 2)
		score_element.setAttributeNS(null, 'style', 'fill:black;;stroke:black;stroke-width:8')
		g.appendChild(score_element)

		var text_element = document.createElementNS(svgNS, "text");
		var text_id = "text_for_" + score_id;
		text_element.setAttributeNS(null, 'id', text_id);
		text_element.setAttributeNS(null, 'text-anchor', 'middle');
		text_element.setAttributeNS(null, 'style', 'fill:white;font-weight: bold; font-size: 20px; dominant-baseline: middle');
		text_element.setAttributeNS(null, 'x', scoreboard_svg_size * score_num + (scoreboard_svg_size / 2))
		text_element.setAttributeNS(null, 'y', scoreboard_svg_size / 2)

		var text_node = document.createTextNode(score);
		text_element.appendChild(text_node);
		g.appendChild(text_element);
	}
}

// Function to update where we sit on the scoring scale
function set_score_position(position)
{
	if (position >= scores_list.length)
		position = scores_list.length - 1
	if (position < 0)
		position = 0

	score_pos = position
	update_scoreboard(position, scores_list)
}

// Function to update the scoreboard highlighting with the current position
function update_scoreboard(position, score_list)
{
	var scoreboard_svg = document.getElementById("score_svg");

	// We also need to make sure all the other scores are unlighted so just spin through the entire list
	for (var score_num = 0; score_num < score_list.length; score_num++)
	{
		var score_id = generate_id_of_object("score_", score_num, score_list.length)
		var score_element = document.getElementById(score_id);

		var score_style = 'fill:black'
		if (position == score_num)
			score_style = score_style + ';stroke:green;stroke-width:' + scoreboard_svg_border
		else
			score_style = score_style + ';stroke:black;stroke-width:' + scoreboard_svg_border
		score_element.setAttributeNS(null, 'style', score_style)
	}
}

// Generate the ID of an object
function generate_id_of_object(name, num, length)
{
	var padding = get_padding_length(length)
	return name + num.toString().padStart(padding, "0")
}


// Determining how much padding we need for a list of objects
function get_padding_length(length)
{
	return Math.floor(Math.log10(length)) + 1
}